### PR TITLE
BI-558 - Improve requiredBy field to support path to root

### PR DESCRIPTION
**Pull request by yahavi**
- [x] All [tests](https://ci.appveyor.com/project/jfrog-ecosystem/build-info) passed. If this feature is not already covered by the tests, I added new tests.
-----

Change the requiredBy field from `List<String>` to `String[][]`.

### CODE REVIEW

File: build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java [hunk 1]
```diff
@@ -161,27 +164,29 @@ public DependencyBuilder remotePath(String remotePath) {
...
      */
-    public DependencyBuilder addRequiredBy(String requiredBy) {
-        if (this.requiredBy == null) {
-            this.requiredBy = new ArrayList<>();
+    public DependencyBuilder addRequiredBy(String[] pathToModuleRoot) {
+        if (requiredBy == null) {
+            requiredBy = new String[][]{pathToModuleRoot};
+        } else {
+            requiredBy = Arrays.copyOf(requiredBy, requiredBy.length + 1);
+            requiredBy[requiredBy.length - 1] = Arrays.copyOf(pathToModuleRoot, pathToModuleRoot.length);
```
Comment (Or-Geva) [inline build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java:188]: We should use this in `build-info-api/src/main/java/org/jfrog/build/api/Dependency.java` to add a new path as well as here

Comment (yahavi) [reply build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java]: Done

File: build-info-api/src/main/java/org/jfrog/build/api/Dependency.java [hunk 2]
```diff
@@ -72,24 +79,38 @@ public void setScopes(Set<String> scopes) {
...
+     * Adds an ID of another dependency required by this one to the required dependencies list
+     *
+     * @param pathToModuleRoot - path from parent dependency to the module ID
+     */
+    public void addRequiredBy(String[] pathToModuleRoot) {
+        if (requiredBy == null) {
+            requiredBy = new String[][]{pathToModuleRoot};
+            return;
+        }
+        requiredBy = Arrays.copyOf(requiredBy, requiredBy.length + 1);
```
Comment (RobiNino) [inline build-info-api/src/main/java/org/jfrog/build/api/Dependency.java:110]: Can maybe replace with [ArrayUtils.add()](https://commons.apache.org/proper/commons-lang/javadocs/api-release/org/apache/commons/lang3/ArrayUtils.html#add-T:A-T-)

File: build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java [hunk 3]
```diff
@@ -161,27 +164,29 @@ public DependencyBuilder remotePath(String remotePath) {
     }
 
     /**
-     * Sets an ID list of other dependencies required by this one
+     * Sets path-to-root of lists of dependencies that directly depend on this dependency.
```
Comment (RobiNino) [inline build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java:167]:
```suggestion
     * Sets path-to-root lists of dependencies that directly depend on this dependency.
```

File: build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java [hunk 4]
```diff
@@ -161,27 +164,29 @@ public DependencyBuilder remotePath(String remotePath) {
...
      *
-     * @param requiredBy Required dependency IDs list
+     * @param requiredBy dependency path-to-root list
```
Comment (RobiNino) [inline build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java:169]:
```suggestion
     * @param requiredBy dependencies path-to-root lists
```

File: build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java [hunk 5]
```diff
@@ -161,27 +164,29 @@ public DependencyBuilder remotePath(String remotePath) {
...
+    public DependencyBuilder requiredBy(String[][] requiredBy) {
         this.requiredBy = requiredBy;
         return this;
     }
 
     /**
      * Adds an ID of another dependency required by this one to the required dependencies list
      *
-     * @param requiredBy Required dependency ID
+     * @param pathToModuleRoot - path from parent dependency to the module ID
```
Comment (RobiNino) [inline build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java:180]:
```suggestion
     * @param pathToModuleRoot - path from parent dependency to the module ID, modules separated.
```

File: build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java [hunk 6]
```diff
@@ -161,27 +164,29 @@ public DependencyBuilder remotePath(String remotePath) {
...
-    public DependencyBuilder addRequiredBy(String requiredBy) {
-        if (this.requiredBy == null) {
-            this.requiredBy = new ArrayList<>();
+    public DependencyBuilder addRequiredBy(String[] pathToModuleRoot) {
+        if (requiredBy == null) {
+            requiredBy = new String[][]{pathToModuleRoot};
+            return this;
         }
-        this.requiredBy.add(requiredBy);
+        requiredBy = Arrays.copyOf(requiredBy, requiredBy.length + 1);
```
Comment (RobiNino) [inline build-info-api/src/main/java/org/jfrog/build/api/builder/DependencyBuilder.java:188]: Can use Apache ArrayUtils here as well if relevant.

File: build-info-api/src/test/java/org/jfrog/build/api/DependencyTest.java [hunk 7]
```diff
@@ -78,6 +79,33 @@ public void testSetters() {
...
+        assertTrue(ArrayUtils.isEmpty(dependency.getRequiredBy()));
+
+        // Add required by A
+        String[] parentA = new String[]{"a", "b", "c"};
+        dependency.addRequiredBy(parentA);
+        assertEquals(ArrayUtils.getLength(dependency.getRequiredBy()), 1);
+        assertEquals(dependency.getRequiredBy()[0], parentA);
+
+        // Add required by B
+        String[] parentB = new String[]{"b", "c", "d"};
```
Comment (RobiNino) [inline build-info-api/src/test/java/org/jfrog/build/api/DependencyTest.java:94]: Test paths of different sizes

File: build-info-api/src/test/java/org/jfrog/build/api/builder/DependencyBuilderTest.java [hunk 8]
```diff
@@ -82,18 +81,22 @@ public void testBuilderSetters() {
...
-        assertEquals(dependency.getProperties().get(propertyKey), propertyValue,
-                "Unexpected dependency property value.");
+        Dependency dependency = new DependencyBuilder().addProperty(propertyKey, propertyValue).build();
+        assertTrue(dependency.getProperties().containsKey(propertyKey), "A dependency property should have been added.");
+        assertEquals(dependency.getProperties().get(propertyKey), propertyValue, "Unexpected dependency property value.");
+    }
+
+    public void testBuilderAddRequiredBy() {
+        String[] requiredByA = {"A", "to", "module", "root", "moduleID"};
+        String[] requiredByB = {"B", "to", "module", "root", "moduleID"};
```
Comment (RobiNino) [inline build-info-api/src/test/java/org/jfrog/build/api/builder/DependencyBuilderTest.java:94]: Test paths of different sizes

**Review by RobiNino — CHANGES_REQUESTED** [review]

**Review by RobiNino — APPROVED** [review]

**Pull request merged**